src/services/refactors/extractSymbol.ts

```diff
@@ -501,6 +501,12 @@ namespace ts.refactor.extractSymbol {
             }
 
         }
+
+        // We could fail to find a scope if
+        //   a) there is no enclosing source file; or
+        //   b) the starting node is the entire source file.

```

Comment:
As for `b`, the types should prevent it from happening: `TargetRange` only uses `Expression | Statement[];`, which can't contain a `SourceFile`.
Aside: It seems wrong to call `firstOrUndefined(range.range)` and then treat it as certainly defined -- I assume this is always non-empty, so it would make more sense to just use `range.range[0]`.

Reply:
Good point about (b) being impossible.

Reply:
I agree about `firstOrUndefined`, but it didn't seem relevant to this change.

Reply:
I think the `firstOrUndefined` call might date from the period when the selection could be empty.  It looks like it would have successfully returned `undefined` if `range.range` was empty.